api/pipeline_spec.proto

```diff
@@ -0,0 +1,420 @@
+      // Resolves an existing artifact.
+      ResolverSpec resolver = 3;
+    }
+  }
+  // Map from executor label to executor spec.
+  map<string, ExecutorSpec> executors = 1;
+}
+
+// Value is the value of the field.
+message Value {
+  oneof value {
+    // An integer value
+    int64 int_value = 1;
+    // A double value
+    double double_value = 2;
+    // A string value
+    string string_value = 3;
+  }
+}
+
+// The definition of a runtime artifact.
+message RuntimeArtifact {
+  // The name of an artifact.
+  string name = 1;
+
+  // The type of the artifact.
+  ArtifactTypeSchema type = 2;
+
+  // The URI of the artifact.
+  string uri = 3;
+
+  // The properties of the artifact.
+  map<string, Value> properties = 4;
+
+  // The custom properties of the artifact.
+  map<string, Value> custom_properties = 5;
+}
+
+// The input of an executor, which includes all the data that
+// can be passed into the executor spec by a string based placeholder.
+//
+// The string based placeholder uses a JSON path to reference to the data
+// in the [ExecutionInput]().
+//
+// `{{$}}`: prints the full [ExecutorInput]() as a JSON string.
+// `{{$.inputs.artifacts['<name>'].uri}}`: prints the URI of an input
+// artifact.
+// `{{$.inputs.artifacts['<name>'].properties['<property name>']}}`: prints
+// the
+//   property of an input artifact.
+// `{{$.inputs.parameters['<name>']}}`: prints the value of an input
+// parameter.
+// `{{$.outputs.artifacts['<name>'].uri}}: prints the URI of an output artifact.
+// `{{$.outputs.artifacts['<name>'].properties['<property name>']}}`: prints the
+//   property of an output artifact.
+// `{{$.outputs.parameters['<name>'].output_file}}`: prints a file path which
+// points to a file and container can write to it to return the value of the
+// parameter..
+// `{{$.outputs.output_file}}`: prints a file path of the output metadata file
+// which is used to send output metadata from executor to orchestrator. The
+// contract of the output metadata is [ExecutorOutput](). When both parameter
+// output file and executor output metadata files are set by the container, the
+// output metadata file will have higher precedence to set output parameters.
+message ExecutorInput {
+  // Message that represents a list of artifacts.
+  message ArtifactList {
+    // A list of parameters.
+    repeated RuntimeArtifact artifacts = 1;
+  }
+
+  // The runtime inputs data of the execution.
+  message Inputs {
+    // Input parameters of the execution.
+    map<string, Value> parameters = 1;
+
+    // Input artifacts of the execution.
+    map<string, ArtifactList> artifacts = 2;
+  }
+
+  // The runtime input artifacts of the task invocation.
+  Inputs inputs = 1;
+
+  // The runtime output parameter.
+  message OutputParameter {
+    // The file path which is used by the executor to pass the parameter value
+    // to the system.
+    string output_file = 1;
+  }
+
+  // The runtime outputs data of the execution.
+  message Outputs {

```

Comment:
In KFP there is a single concept of an output (no discrimination between output parameters and output artifacts).
How do you plan to map KFP pipelines to this schema?

Reply:
We have an internal discussion about this already. Let's discuss this offline.